### PR TITLE
Add MonadPropT to master

### PR DESCRIPTION
**Pull request by euisuny**
Porting over `PropT` and relevant proofs from vellvm (https://github.com/vellvm/vellvm/blob/dev/src/coq/Utils/PropT.v)

### CODE REVIEW

File: theories/Basics/MonadPropT.v [hunk 1]
```diff
@@ -0,0 +1,1496 @@
...
+Lemma bind_bind_counterexample:
+    exists t, bind PA (fun a => bind (KB a) KC) t /\ ~ (bind (bind PA KB) KC t).
+Proof.
+  exists t.
+  split.
+  apply bind_right_assoc.
+  apply not_bind_left_assoc.
+Qed.
+
+End BIND_BIND_COUNTEREXAMPLE.
```
[the remark this answers is not in the document]
Comment (euisuny) [reply theories/Basics/MonadPropT.v:1496]: Yeah, it doesn't satisfy all the monad laws.

[the remark this answers is not in the document]
Comment (Lysxia) [reply theories/Basics/MonadPropT.v:1496]: oh haha I was just commenting on the number of lines. Is it not a problem that it's not lawful though?

[the remark this answers is not in the document]
Comment (euisuny) [reply theories/Basics/MonadPropT.v:1496]: It's a problem, but a known problem. We don't need/use the lawfulness, at least in the Vellvm proofs.

**Comment by Lysxia** [comment]
I wonder if `bind` could be made lawful using an ad hoc "bind-like" relation, instead of using `bind` and `eutt`. But I'm not going to let that block this PR.

**Pull request merged**